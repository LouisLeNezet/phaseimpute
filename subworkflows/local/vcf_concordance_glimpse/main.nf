include { GLIMPSE2_CONCORDANCE        } from '../../../modules/nf-core/glimpse2/concordance'
include { CONCATENATE                 } from '../../../modules/local/concatenate'
include { ADD_COLUMNS                 } from '../../../modules/local/addcolumns'
include { GUNZIP                      } from '../../../modules/nf-core/gunzip'

workflow VCF_CONCORDANCE_GLIMPSE {

    take:
        ch_vcf_emul   // VCF file with imputed genotypes [[id, chr, region, panel, simulate, tools], vcf, csi]
        ch_vcf_truth  // VCF file with truth genotypes   [[id, chr, region], vcf, csi]
        ch_vcf_freq   // VCF file with panel frequencies [[panel, chr], vcf, csi]

    main:

    ch_versions = Channel.empty()

    ch_concordance = ch_vcf_emul
        .map{
            metaICRPST, vcf, csi ->
            [metaICRPST.subMap(["id", "chr", "region", "panel"]), metaICRPST, vcf, csi]
        }
        .combine(ch_vcf_truth, by:0)
        .map{metaICRP, metaIPCRTS, emul, e_csi, truth, t_csi ->
            [metaICRP.subMap(["chr"]), metaIPCRTS, emul, e_csi, truth, t_csi]
        }
        .combine(ch_vcf_freq.map{metaCRP, vcf, csi ->
                    [metaCRP.subMap(["chr"]), metaCRP, vcf, csi]},
                by:0)
        .map{metaC, metaIPCRTS, emul, e_csi, truth, t_csi, metaCRP, freq, f_csi ->
            [metaIPCRTS, emul, e_csi, truth, t_csi, freq, f_csi, [], metaIPCRTS.region]
        }

    GLIMPSE2_CONCORDANCE (
        ch_concordance,
        [[], [], "0 0.01 0.05 0.1 0.2 0.5", [], []],
        0.9, 5
    )
    GUNZIP(GLIMPSE2_CONCORDANCE.out.errors_grp)
    ADD_COLUMNS(GUNZIP.out.gunzip)

    CONCATENATE(ADD_COLUMNS.out.txt
                    .map{meta, txt -> [["id":"TestQuality"], txt]}
                    .groupTuple()
    )


    emit:
    stats     = CONCATENATE.out.txt         // [ meta, txt ]
    versions  = ch_versions                 // channel: [ versions.yml ]
}
