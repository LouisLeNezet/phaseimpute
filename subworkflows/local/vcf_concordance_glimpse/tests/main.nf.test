nextflow_workflow {

    name "Test Subworkflow VCF_CONCORDANCE_GLIMPSE"
    script "../main.nf"
    config "./nextflow.config"

    workflow "VCF_CONCORDANCE_GLIMPSE"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/vcf_concordance_glimpse"
    tag "vcf_concordance_glimpse"

    tag "bcftools"
    tag "bcftools/index"
    tag "glimpse"
    tag "glimpse/phase"
    tag "glimpse/concordance"

    test("vcf_concordance_glimpse") {
        setup {
            run("GLIMPSE_PHASE") {
                script "../../../../modules/nf-core/glimpse/phase/main.nf"
                process {
                """
                ch_sample     = Channel.of('NA12878 2', 'NA12878_2 2').collectFile(name: 'sampleinfos.txt', newLine: true)
                region        = Channel.fromList([
                    ["chr21:16600000-16750000","chr21:16650000-16700000"]
                ])
                input_vcf     = Channel.fromList([
                    [[ id:'NA12878', chr:'21', region:'chr21:16650000-16700000', panel: '1000GP', depth:'1', tools: 'Glimpse'], // meta map
                    file(params.modules_testdata_base_path + "delete_me/glimpse/NA12878.chr21.s.1x.vcf.gz", checkIfExists: true),
                    file(params.modules_testdata_base_path + "delete_me/glimpse/NA12878.chr21.s.1x.vcf.gz.csi", checkIfExists: true)],
                    [[ id:'NA12878_2', chr:'21', region:'chr21:16650000-16700000', panel: '1000GP', depth:'0.5', tools: 'Glimpse2'], // meta map
                    file(params.modules_testdata_base_path + "delete_me/glimpse/NA12878.chr21.s.1x.vcf.gz", checkIfExists: true),
                    file(params.modules_testdata_base_path + "delete_me/glimpse/NA12878.chr21.s.1x.vcf.gz.csi", checkIfExists: true)]
                ])
                ref_panel     = Channel.of([
                    file(params.modules_testdata_base_path + "delete_me/glimpse/1000GP.chr21.noNA12878.s.bcf", checkIfExists: true),
                    file(params.modules_testdata_base_path + "delete_me/glimpse/1000GP.chr21.noNA12878.s.bcf.csi", checkIfExists: true)
                ])
                ch_map        = Channel.of([
                    file(params.modules_testdata_base_path + "delete_me/glimpse/chr21.b38.gmap.gz", checkIfExists: true),
                ])

                input[0]      = input_vcf
                    | combine(ch_sample)
                    | combine(region)
                    | combine(ref_panel)
                    | combine(ch_map)
                """
                }
            }
            run("BCFTOOLS_INDEX") {
                script "../../../../modules/nf-core/bcftools/index/main.nf"
                process {
                """
                input[0] = GLIMPSE_PHASE.out.phased_variants
                """
                }
            }
        }
        when {
            workflow {
                """

                allele_freq = Channel.of([
                    [panel:'1000GP', chr:'21'], // meta map
                    file(params.modules_testdata_base_path + "delete_me/glimpse/1000GP.chr21.noNA12878.s.sites.vcf.gz",checkIfExists:true),
                    file(params.modules_testdata_base_path + "delete_me/glimpse/1000GP.chr21.noNA12878.s.sites.vcf.gz.csi",checkIfExists:true)
                ])
                truth = Channel.fromList([
                    [[id:'NA12878', chr:'21', region:'chr21:16650000-16700000'], // meta map
                    file(params.modules_testdata_base_path + "delete_me/glimpse/NA12878.chr21.s.bcf",checkIfExists:true),
                    file(params.modules_testdata_base_path + "delete_me/glimpse/NA12878.chr21.s.bcf.csi",checkIfExists:true)],
                    [[id:'NA12878_2', chr:'21', region:'chr21:16650000-16700000'], // meta map
                    file(params.modules_testdata_base_path + "delete_me/glimpse/NA12878.chr21.s.bcf",checkIfExists:true),
                    file(params.modules_testdata_base_path + "delete_me/glimpse/NA12878.chr21.s.bcf.csi",checkIfExists:true)]
                ])
                estimate = GLIMPSE_PHASE.out.phased_variants
                    | join (BCFTOOLS_INDEX.out.csi)
                input[0] = estimate
                input[1] = truth
                input[2] = allele_freq
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }

    }
}
