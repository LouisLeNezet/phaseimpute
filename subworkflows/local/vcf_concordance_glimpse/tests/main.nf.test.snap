{
    "vcf_concordance_glimpse": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "TestQuality"
                        },
                        "TestQuality.txt:md5,910b294df62dbe64e8f16379428d93ad"
                    ]
                ],
                "1": [
                    
                ],
                "stats": [
                    [
                        {
                            "id": "TestQuality"
                        },
                        "TestQuality.txt:md5,910b294df62dbe64e8f16379428d93ad"
                    ]
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-25T12:44:54.967846019"
    }
}